api/envoy/config/api_listener/v2/api_listener.proto

```diff
@@ -0,0 +1,26 @@
+syntax = "proto3";
+
+package envoy.config.api_listener.v2;
+
+option java_outer_classname = "ApiListenerProto";
+option java_multiple_files = true;
+option java_package = "io.envoyproxy.envoy.config.api_listener.v2";
+
+option java_generic_services = true;
+
+import "envoy/config/filter/network/http_connection_manager/v2/http_connection_manager.proto";
+
+import "gogoproto/gogo.proto";
+
+option (gogoproto.equal_all) = true;
+
+// [#not-implemented-hide:]
+// Describes a type of API listener, which is used in non-proxy clients. The type of API
+// exposed to the non-proxy application depends on the type of API listener.
+message ApiListener {
+  oneof api_type {
```

Comment:
Can this just be a `Filter`? then you don't have to explicitly use HCM so build will pass, also we don't need to modify this to support other protocols.

Reply:
The goal here is that this message should explicitly allow only the specific type of configuration appropriate to the type of API listener.  If we make it a `Filter`, we allow invalid configs like specifying a type of filter that is not related to a listener; this way, that kind of invalid config is structurally impossible, which makes validation easier.

Reply:
Don't we need to validate anyway (with PGV and additional checks to check if the config is valid HCM config), so just one more type check shouldn't be a deal breaker, no?

Reply:
I'll let @htuch and @mattklein123 weigh in on this.

From my perspective, the fact that the HCM config happens to be a filter config is basically a coincidence; in the future, we could have other types of API listeners whose fields are not existing filter messages.  And the vast majority of filter messages are not suitable as configuration for a type of API listener.  So it seems better to me to tailor this to what we actually need rather than providing unnecessary flexibility that IMHO makes the configuration harder to understand.

Reply:
The Filter message is just a name with Any or Struct, what do you mean by "the vast majority of filter messages are not suitable as configuration for a type of API listener"?

Reply:
What I mean is that a `Filter` field expects its `Any` field to be populated with one of the messages in the envoy/config/filter tree (or a message for a third-party filter).  The vast majority of those messages will never be suitable for use as a type of API listener.

The concept here is that an API listener exposes a type-specific API to the application.  The HTTP listener conceptually provides an API that exposes reading and writing headers, data, and trailers for an HTTP request; this happens to map to the functionality provided by the HCM, and the HCM happens to also be used as a filter in Envoy, but that's just a coincidence.  In the future, we could also have (e.g.) an SQL API listener that exposes an API based around sending SQL queries and processing result records, or any other L7 protocol for which there is no existing Envoy filter.  And conversely, most of the existing filters do not define L7 protocols and therefore can't act as API listeners; for example, it would be inherently non-sensical to say that we are going to expose an API for applications to access a service via (e.g.) fault injection or access logging.

Reply:
> What I mean is that a `Filter` field expects its `Any` field to be populated with one of the messages in the envoy/config/filter tree (or a message for a third-party filter).  The vast majority of those messages will never be suitable for use as a type of API listener.

That's why I said, "do a type check and reject the config as DPLB", seems you just don't like the name "Filter"?

>  In the future, we could also have (e.g.) an SQL API listener that exposes an API based around sending SQL queries and processing result records, or any other L7 protocol for which there is no existing Envoy filter. And conversely, most of the existing filters do not define L7 protocols and therefore can't act as API listeners; for example, it would be inherently non-sensical to say that we are going to expose an API for applications to access a service via (e.g.) fault injection or access logging.

This doesn't seems conflicting with use of Any (or "Filter"), it is just a free form Any, the xDS protocol we're making here should guard that at reporting capabilities of which type of Any is supported, IMHO that should not be limited by the types of proto message level here.

Reply:
I do think that using `Filter` is worse than `Any`, because it implies to the reader that any filter configuration can be used here, which is not the case.  API listeners really have nothing to do with filters, and IMHO it does not make sense to conflate two unrelated concepts.

With regard to the difference between a oneof and an `Any`, I think the trade-off is that an `Any` makes it easier to add new types, whereas a oneof provides more type safety and structurally prevents invalid configurations.  I realize that we can also catch that kind of problem with validation, and I'm certainly not opposed to validation, but I think it makes everyone's lives easier if the underlying structure doesn't allow misconfiguration in the first place.  If that weren't the case, every single field in every proto would be an `Any`. :)

IMHO, `Any` makes sense only in cases where we expect new types to be added frequently, especially by third parties.  But I don't think that's the case here; I don't expect that adding new types of API listener will happen that often, so the additional flexibility of using `Any` isn't much of an offsetting advantage here.

In any case, as mentioned below, we have no choice but to go with an `Any` for now, due to the go proto circular dependency issue.  But I've added a note that we should convert this to a oneof in the v3 API.

Reply:
I agree with Mark here that makes sense to be more specific, but given that we can't due to circular dependency issues, it seems fine to use Any for now.

Reply:
> IMHO, `Any` makes sense only in cases where we expect new types to be added frequently, especially by third parties.  But I don't think that's the case here; I don't expect that adding new types of API listener will happen that often, so the additional flexibility of using `Any` isn't much of an offsetting advantage here.

If we don't expect third parties add new types here then oneof is totally fine. If we don't expect any "I have my own API library want to implement in-house RPC protocol with UDPA" to be come up, this is fine. Historically we did convert oneof to Any or add Any to oneof in several places in Envoy (cluster, retry policies...), so if we're fine on that expectation I'm good with that.